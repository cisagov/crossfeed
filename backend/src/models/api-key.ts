import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
