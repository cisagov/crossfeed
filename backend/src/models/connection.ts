import { createConnection, Connection } from 'typeorm';
import {
  Domain,
  Service,
  Vulnerability,
  Scan,
  Organization,
  User,
  Role,
  ScanTask
} from '.';

let connection: Connection | null = null;

const connectDb = async (logging?: boolean) => {
  const connection = createConnection({
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT ?? ''),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    entities: [
      Domain,
      Service,
      Vulnerability,
      Scan,
      Organization,
      User,
      Role,
      ScanTask
    ],
    synchronize: false,
    name: 'default',
    dropSchema: false,
    logging: logging ?? false
  });
  return connection;
};

export const connectToDatabase = async (logging?: boolean) => {
  if (!connection?.isConnected) {
    connection = await connectDb(logging);
  }
  if (typeof jest === 'undefined') {
    console.log('=> DB Connected');
  }
  return connection;
};
