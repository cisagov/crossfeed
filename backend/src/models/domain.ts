import {
  Entity,
  Index,
  Column,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  BeforeInsert,
  ManyToOne
} from 'typeorm';
import { Service } from './service';
import { Organization } from './organization';
import { Vulnerability } from './vulnerability';
import { Scan } from './scan';
import { Webpage } from './webpage';

@Entity()
@Index(['name', 'organization'], { unique: true })
export class Domain extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  /** When this model was last synced with Elasticsearch. */
  @Column({
    type: 'timestamp',
    nullable: true
  })
  syncedAt: Date | null;

  @Column({
    nullable: true
  })
  ip: string;

  /** Associated root domain that led to the discovery of this domain. */
  @Column({
    nullable: true
  })
  fromRootDomain: string;

  @ManyToOne((type) => Scan, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  discoveredBy: Scan;

  @Column({
    length: 512
  })
  reverseName: string;

  @Column({
    length: 512
  })
  name: string;

  @OneToMany((type) => Service, (service) => service.domain)
  services: Service[];

  @OneToMany((type) => Vulnerability, (vulnerability) => vulnerability.domain)
  vulnerabilities: Vulnerability[];

  @OneToMany((type) => Webpage, (webpage) => webpage.domain)
  webpages: Service[];

  @ManyToOne((type) => Organization, { onDelete: 'CASCADE', nullable: false })
  organization: Organization;

  @Column({
    length: 512,
    nullable: true,
    type: 'varchar'
  })
  screenshot: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  country: string | null;

  @Column({
    nullable: true,
    type: 'varchar'
  })
  asn: string | null;

  @Column({
    default: false
  })
  cloudHosted: boolean;

  /** SSL Certificate information  */
  @Column({
    type: 'jsonb',
    nullable: true
  })
  ssl: {
    issuerOrg: string;
    issuerCN: string;
    validFrom: string;
    validTo: string;
    protocol: string;
    altNames: string;
    bits: string;
    fingerprint: string;
  } | null;

  /** Censys Certificates results */
  @Column({
    type: 'jsonb',
    default: {}
  })
  censysCertificatesResults: {
    [x: string]: any;
  };

  @BeforeInsert()
  setLowerCase() {
    this.name = this.name.toLowerCase();
  }

  @BeforeInsert()
  setReverseName() {
    this.reverseName = this.name.split('.').reverse().join('.');
  }
}
