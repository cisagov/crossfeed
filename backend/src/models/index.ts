export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
