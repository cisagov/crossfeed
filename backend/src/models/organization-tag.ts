import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToMany,
  JoinTable,
  Column
} from 'typeorm';
import { Organization, Scan } from '.';

@Entity()
@Index(['name'], { unique: true })
export class OrganizationTag extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  /**
   * Organizations that are labeled with this tag
   */
  @ManyToMany((type) => Organization, (organization) => organization.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  organizations: Organization[];

  /**
   * Scans that have this tag enabled, and will run against all tagged organizations
   */
  @ManyToMany((type) => Scan, (scan) => scan.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  scans: Scan[];
}
