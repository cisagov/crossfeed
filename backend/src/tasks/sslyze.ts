import { CommandOptions } from './ecs-client';
import { getLiveWebsites, LiveDomain } from './helpers/getLiveWebsites';
import PQueue from 'p-queue';
import { execSync } from 'child_process';

const sslyze = async (domain: LiveDomain): Promise<void> => {
  try {
    const path = `/tmp/${domain.name}.crt`;
    const res = execSync(
      `echo "" | openssl s_client -connect ${domain.name}:443 > ${path}; openssl x509 -in ${path} -noout -enddate`,
      {
        shell: '/bin/bash'
      }
    );
    const output = String(res);
    const result = output.match(/notAfter=(.*)/);
    if (!result || !result[1]) {
      console.error('No notAfter date found, output is', output);
      return;
    }
    domain.ssl = {
      ...(domain.ssl || {}),
      validTo: new Date(result![1]).toISOString()
    };
    await domain.save();
  } catch (e) {
    console.error(domain.name, e);
  }
};

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running sslyze on organization', organizationName);

  const liveWebsites = await getLiveWebsites(organizationId!, [], true);
  const queue = new PQueue({ concurrency: 5 });
  await Promise.all(liveWebsites.map((site) => queue.add(() => sslyze(site))));

  console.log(`sslyze finished for ${liveWebsites.length} domains`);
};
