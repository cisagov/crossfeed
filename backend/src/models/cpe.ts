import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  Unique
} from 'typeorm';
import { Cve } from './cve';

@Entity()
@Unique(['name', 'version', 'vendor'])
export class Cpe extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column()
  version: string;

  @Column()
  vendor: string;

  @Column()
  lastSeenAt: Date;

  @ManyToMany(() => Cve, (cve) => cve.cpes)
  cves: Cve[];
}
