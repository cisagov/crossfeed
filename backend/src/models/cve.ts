import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToMany,
  BaseEntity,
  JoinTable,
  Unique
} from 'typeorm';
import { Cpe } from './cpe';

//TODO: Refactor column names to camelCase to match the rest of the codebase?
@Entity()
@Unique(['name'])
export class Cve extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ nullable: true })
  name: string;

  @Column({ nullable: true })
  publishedAt: Date;

  @Column({ nullable: true })
  modifiedAt: Date;

  @Column({ nullable: true })
  status: string;

  @Column({ nullable: true })
  description: string;

  @Column({ nullable: true })
  cvssV2Source: string;

  @Column({ nullable: true })
  cvssV2Type: string;

  @Column({ nullable: true })
  cvssV2Version: string;

  @Column({ nullable: true })
  cvssV2VectorString: string;

  @Column({ nullable: true })
  cvssV2BaseScore: string;

  @Column({ nullable: true })
  cvssV2BaseSeverity: string;

  @Column({ nullable: true })
  cvssV2ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV2ImpactScore: string;

  @Column({ nullable: true })
  cvssV3Source: string;

  @Column({ nullable: true })
  cvssV3Type: string;

  @Column({ nullable: true })
  cvssV3Version: string;

  @Column({ nullable: true })
  cvssV3VectorString: string;

  @Column({ nullable: true })
  cvssV3BaseScore: string;

  @Column({ nullable: true })
  cvssV3BaseSeverity: string;

  @Column({ nullable: true })
  cvssV3ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV3ImpactScore: string;

  @Column({ nullable: true })
  cvssV4Source: string;

  @Column({ nullable: true })
  cvssV4Type: string;

  @Column({ nullable: true })
  cvssV4Version: string;

  @Column({ nullable: true })
  cvssV4VectorString: string;

  @Column({ nullable: true })
  cvssV4BaseScore: string;

  @Column({ nullable: true })
  cvssV4BaseSeverity: string;

  @Column({ nullable: true })
  cvssV4ExploitabilityScore: string;

  @Column({ nullable: true })
  cvssV4ImpactScore: string;

  @Column('simple-array', { nullable: true })
  weaknesses: string[];

  @Column('simple-array', { nullable: true })
  references: string[];

  @ManyToMany(() => Cpe, (cpe) => cpe.cves, {
    cascade: true
  })
  @JoinTable()
  cpes: Cpe[];
}
