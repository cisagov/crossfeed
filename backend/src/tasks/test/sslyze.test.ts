import { handler as sslyze } from '../sslyze';
import { connectToDatabase, Organization, Service, Domain } from '../../models';

jest.mock('child_process', () => ({
  execSync: function (a, b) {
    expect([a, b]).toMatchSnapshot();
    return `depth=2 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert Global Root CA
  verify return:1
  depth=1 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert Secure Site ECC CA-1
  verify return:1
  depth=0 C = US, ST = District Of Columbia, L = Washington, O = Department of Homeland Security, CN = www3.dhs.gov
  verify return:1
  DONE
  notAfter=Mar  7 12:00:00 2021 GMT`;
  }
}));

describe('sslyze', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  let organization;
  beforeEach(async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });
  test('basic test', async () => {
    let domain = await Domain.create({
      organization,
      name: 'www.cisa.gov',
      ip: '0.0.0.0'
    }).save();
    const service = await Service.create({
      service: 'https',
      port: 443,
      domain
    }).save();
    await sslyze({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });
    domain = (await Domain.findOne(domain.id)) as Domain;
    expect(domain.ssl).toMatchInlineSnapshot(`
      Object {
        "validTo": "2021-03-07T12:00:00.000Z",
      }
    `);
  });
  test('should not be called on non-https domains', async () => {
    const domain = await Domain.create({
      organization,
      name: 'www.cisa.gov',
      ip: '0.0.0.0'
    }).save();
    const service = await Service.create({
      service: 'http',
      port: 80,
      domain
    }).save();
    await sslyze({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });
  });
});
