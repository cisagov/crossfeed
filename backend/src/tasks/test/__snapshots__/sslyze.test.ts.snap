// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sslyze basic test 1`] = `
Array [
  "echo \\"\\" | openssl s_client -connect www.cisa.gov:443 > /tmp/www.cisa.gov.crt; openssl x509 -in /tmp/www.cisa.gov.crt -noout -enddate",
  Object {
    "shell": "/bin/bash",
  },
]
`;
